File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.cpp/core/src/impl/CommandStackImpl.cpp
#include "CommandStackImpl.h"

#include <stdexcept>

namespace commands {

CommandStackImpl::CommandStackImpl() {
}

void CommandStackImpl::execute(std::shared_ptr<commands::Command> command) {
    addToStack(command);
	command->execute();
}

void CommandStackImpl::undo() {
    auto stack = getStack();
    if (stack.empty()) {
        throw std::runtime_error("cannot undo with empty stack");
    }

    auto command = stack.back();
    removeFromStack(command);
    command->undo();
}

void CommandStackImpl::redo() {
	throw std::runtime_error("not implemented");
}

void CommandStackImpl::undoAll() {
    while (!getStack().empty()) {
        undo();
    }
}

void CommandStackImpl::redoAll() {
	throw std::runtime_error("not implemented");
}

}

File: de.unistuttgart.iste.sqa.mpw.karasimulator.java/core/src/main/java/de/unistuttgart/kara/commands/impl/CommandStackImpl.java
package de.unistuttgart.kara.commands.impl;

import de.unistuttgart.kara.commands.CommandStack;
import de.unistuttgart.kara.commands.Command;

import static de.unistuttgart.iste.rss.utils.Preconditions.checkState;

public class CommandStackImpl extends CommandStack {

	public CommandStackImpl() {
		super();
	}


	@Override
	public void execute(Command command) {
		command.execute();
		addToExecutedCommands(command);
	}

	@Override
	public synchronized void undo() {
		var stack = getExecutedCommands();
		checkState(stack.size() > 0, "Cannot undo");

		final Command undoneCommand = stack.remove(stack.size() - 1);
		undoneCommand.undo();
		getUndoneCommands().add(undoneCommand);
	}

	@Override
	public synchronized void redo() {
		var undoStack = getUndoneCommands();
		checkState(undoStack.size() > 0, "Cannot redo");

		final Command undoneCommand = undoStack.remove(undoStack.size() - 1);
		undoneCommand.redo();
		getExecutedCommands().add(undoneCommand);
	}

	@Override
	public synchronized void undoAll() {
		var stack = getExecutedCommands();
		while (stack.size() > 0) {
			undo();
		}
	}

	@Override
	public synchronized void redoAll() {
		var undoStack = getUndoneCommands();
		while (undoStack.size() > 0) {
			redo();
		}
	}
}
